common/riff.ksy

```diff
@@ -0,0 +1,90 @@
+meta:
+  id: riff
+  title: Resource Interchange File Format (RIFF)
+  license: CC0-1.0
+  endian: le
+  encoding: ASCII
+  xref:
+    justsolve: RIFF
+    loc: fdd000025
+    wikidata: Q1196805
+doc: |
+  The Resource Interchange File Format (RIFF) is a generic file container format
+  for storing data in tagged chunks. It is primarily used to store multimedia
+  such as sound and video, though it may also be used to store any arbitrary data.
+
+  The Microsoft implementation is mostly known through container formats
+  like AVI, ANI and WAV, which use RIFF as their basis.
+seq:
+  - id: riff_chunk
+    type: chunk
+types:
+  parent_chunk_data:
+    seq:
+      - id: form_type
+        type: u4be
```

Comment:
how about `le` or even a fixed-length `str` here? IMHO there is no need in a enum. For individual formats we need a way to require a chunk to be of a certain type, but it is currently impossible, needs a fix in KSC, probably even more than 1.

So I guess the following architecture can be beneficial:
1. separate files for chunkspecs in a `riff` subdir of the dir where the target format spec lies.
2. a `riff` spec importing them all and having an optional param of allowed chunk type. Ideally it should assert than chunk type matches, but for now I suggest to just check that and if it doesn't, then don't parse using `if`.
2. a spec file for formats importing `riff` spec and accepting the arg which

for example the following tree
```yaml
repo_root:
  - media:
    - riff: 
      - wave_riff_chunk
      - umid_riff_chunk
    - riff.ksy
    - wav.ksy
    - midi.ksy
  - image
    - riff:
      - cdr_riff_chunk 
    - corel_draw_cdr.ksy
```

Reply:
> how about `le` or even a fixed-length `str` here? IMHO there is no need in a enum.

Yes, fixed-length `str` will be probably better. I thought that it's the recommended approach for four-character codes to be represented as integers because of [the user guide](https://doc.kaitai.io/user_guide.html#_fourcc), but here it doesn't make much sense. On top of that, strings are much more readable.


> So I guess the following architecture can be beneficial: (...)

If I understand what you're suggesting:
```YAML
# /media/riff.ksy
meta:
  id: riff
  endian: le
  imports:
    - media/riff/wav_riff_chunk
    - media/riff/rmid_riff_chunk
params:
  - id: allowed_chunk_type
    type: str
seq:
  - id: chunk_id
    contents: RIFF
  - id: chunk_size
    type: u4
  - id: chunk_data
    type:
      switch-on: allowed_chunk_type
      cases:
        '"WAVE"': wav_riff_chunk
        '"RMID"': rmid_riff_chunk
        _: unknown_chunk
    size: chunk_size
  - id: pad_byte
    size: chunk_size % 2
types:
  unknown_chunk: {}
```
```YAML
# /media/riff/wav_riff_chunk.ksy
meta:
  id: wav_riff_chunk
  endian: le
seq:
  - id: form_type
    contents: WAVE
  - id: subchunks
    type: chunk_type
    repeat: eos
types:
  # copy & paste `chunk_type` definition from `media/wav.ksy` 
  # and its dependencies here (those >500 lines wouldn't fit here 😃)
```
```YAML
# /media/riff/rmid_riff_chunk.ksy
meta:
  id: rmid_riff_chunk
  encoding: ASCII
  endian: le
  imports:
    - media/standard_midi_file
  doc: |
    A RIFF wrapper around SMF.
seq:
  - id: form_type
    contents: RMID
  - id: subchunks
    type: subchunk
    repeat: eos
types:
  subchunk:
    seq:
      - id: chunk_id
        type: str
        size: 4
        pad-right: 0x20
      - id: chunk_size
        type: u4
      - id: chunk_data
        type:
          switch-on: chunk_id
          cases:
            '"data"': standard_midi_file
        size: chunk_size
      - id: pad_byte
        size: chunk_size % 2
```
```YAML
# /media/wav.ksy
meta:
  id: wav
  imports:
    - media/riff
seq:
  - id: riff_chunk
    type: riff('WAVE')
```
```YAML
# /media/rmid.ksy
meta:
  id: rmid
  imports:
    - media/riff
seq:
  - id: riff_chunk
    type: riff('RMID')
```

I don't know. It's definitely some improvement, but I'm not quite satisfied with it.

1. We're basically moving all chunkspecs to `riff/*_riff_chunk.ksy` files, so the particular format specs become just empty wrappers around `/media/riff.ksy`.
2. Imagine that there are dozens of chunkspecs in `riff/` folder. RIFF spec `/media/riff.ksy` imports all of them, but only one at a time is used. It adds many unnecessary dependencies.
3. There is still quite a lot of repetition. It is necessary to define basic RIFF `chunk` and `parent_chunk_data` over and over again.

Yet it is IMHO the best thing what we can do in KS right now. If we had generics (kaitai-io/kaitai_struct#135) and validation (kaitai-io/kaitai_struct#435 - we already have, but not in all target languages), we would have been able to do this (syntax from https://github.com/kaitai-io/kaitai_struct/issues/273#issue-260926722):
```YAML
# /common/riff/chunk.ksy
meta:
  id: chunk
  encoding: ASCII
  endian: le
templates:
  - expected_chunk_id
  - data_generic
seq:
  - id: chunk_id
    type: str
    valid: expected_chunk_id
    size: 4
    pad-right: 0x20
  - id: chunk_size
    type: u4
  - id: chunk_data
    type: data_generic
    size: chunk_size
  - id: pad_byte
    size: chunk_size % 2
```
```YAML
# /common/riff/parent_chunk_data.ksy
meta:
  id: parent_chunk_data
  encoding: ASCII
  endian: le
templates:
  - expected_form_type
  - subchunk_generic
seq:
  - id: form_type
    type: str
    valid: expected_form_type
    size: 4
    pad-right: 0x20
  - id: subchunks
    type: subchunk_generic
    repeat: eos
```
```YAML
# /common/wav.ksy
meta:
  id: wav
  imports:
    - common/riff/chunk
    - common/riff/parent_chunk_data
seq:
  - id: riff_chunk
    type: >-
      chunk<
        'RIFF',
        parent_chunk_data<
          'WAVE',
          chunk<
            _, # somehow omit the `expected_chunk_id` field
            chunk_data
          >
        >
      >
types:
  chunk_data:
    seq:
      - id: data
        type:
          switch-on: _parent.as<chunk>.chunk_id
          cases:
            '"fmt"': format_chunk_data
            '"bext"': bext_chunk_data
            '"cue"': cue_chunk_data
            '"data"': data_chunk_data
            '"LIST"': >-
              parent_chunk_data<
                _, # omit the `expected_form_type` field
                list_subchunk
              >
  list_subchunk:
    seq:
      id: data
      type:
          switch-on: _parent.as<parent_chunk_data>.form_type
          cases:
            '"INFO"': strz
# ...
```
I like the brevity, simplicity and clarity so much! I can help with the implementation in KSC if you want.